File: Python/NEST.py
import numpy as np

from checkArgs import check_args
from pvalFun import pval_fun
from enrichScore import enrichScore
from statFun_lm import stat_fun_lm
from statFun_gam_deltaRsq import statFun_gam_deltaRsq


def NEST(statFun,args, net_maps, one_sided=True, seed=None, statFun_custom=None):
    '''
    Input:
        statFun:
        args: 
        network_maps: 

    Output:
        pval: p-value
        ES: observed enrichment score
        ES_perm: enrichment score for permutations
        running_sum: running sum for observed statistics
    '''
    if not (isinstance(net_maps, list) or (isinstance(net_maps, np.ndarray) and net_maps.ndim == 1)):
        print("net_maps argument should be a list, of 1-d numpy array")
        return None      
    if args['X'].shape[1] != len(net_maps):
        print("The network should be specified as a vector with length equal to number of columns of X.")
        return None
    if not isinstance(net_maps, np.ndarray):
        net_maps = np.array(net_maps)
    is_binary = np.all(np.isin(net_maps, [0, 1]))
    if not is_binary:
        print("net.maps should include only 0's and 1's (1 = in network/ROI; 0 = outside network/ROI)")
        return None
    
    if statFun == "lm":
        required_args = ["X", "y"]
        optional_args = ["Z", "type", "FL", "getNull", "n_perm"]

        args = check_args(args, required_args, optional_args)

        if args:

            statFun_out = stat_fun_lm(X=args['X'], y=args['y'], Z=args['Z'], type=args['type'],
                                     seed=seed, FL=args['FL'], n_perm=args['n_perm'],
                                     getNull=args['getNull'])
        else:
            print('Error. Please check args.')
            return None
    elif statFun == 'gam_mnwald':
        # to be added
        print('No implementation.')
        return None
    
    elif statFun == 'gam_deltaRsq':
        # added deltaRsq according to the R implementation (Credit for Autrey Luo)
        required_args = ['X',"dat",'gam_full_smoother','gam_null_smoother','y_in_gam','y_in_lm','y_permute']
        optional_args = ["getNull", "n_perm"]
        args = check_args(args, required_args, optional_args)   
        if args:
            statFun_out = statFun_gam_deltaRsq(
                                X=args['X'], 
                                dat=args['dat'], 
                                gam_full_smoother=args['gam_full_smoother'], 
                                gam_null_smoother=args['gam_null_smoother'], 
                                y_in_gam=args['y_in_gam'],
                                y_in_lm=args['y_in_lm'], 
                                y_permute=args['y_permute'], 
                                seed=seed, 
                                n_perm=args['n_perm'], 
                            )
            print(statFun_out['T_obs'])
            print(len(statFun_out['T_obs']))

        else:
            print('Error. Please check args.')
            return None     
    
    # For custom methods, the user should at least specify the arguments X, y, and statFun_custom (method name). It is optional for users to add more arguments to args as needed.
    elif statFun == 'custom':
        required_args = ["X", "y"]
        optional_args = []
        #optional_args = ["Z", "type", "FL", "getNull", "n_perm"]
        args = check_args(args, required_args, optional_args)

        if args:
            statFun_out = statFun_custom(**args)
    else:
        return None

    networks = ['net_'+str(x) for x in range(1)]

    ES_obs,running_sum = enrichScore(statFun_out['T_obs'],net_maps,networks,one_sided)

    ES_null = {}
    for i in range(args['n_perm']):
        ES_perm, _ = enrichScore(statFun_out['T_null'][i],net_maps,networks,one_sided)
        for net, val in ES_perm.items():
            if net in ES_null:
                ES_null[net].append(val)
            else:
                ES_null[net]=[val]


    for net in networks:
        pval = pval_fun(ES_obs[net],ES_null[net])


    return pval,ES_obs,ES_null,running_sum

File: R/NEST.R
#' NEST function
#'@importFrom parallel mclapply
#'@param statFun A string specifying the statistical method to use. Options include:
#'   \describe{
#'     \item{"lm"}{Linear regression model (`statFun.lm`).}
#'     \item{"gam.mvwald"}{Generalized additive model with multivariate Wald tests (`statFun.gam.mvwald`).}
#'     \item{"gam.deltaRsq"}{Generalized additive model for delta adjusted R-squared (`statFun.gam.deltaRsq`).}
#'     \item{"custom"}{Custom statistical function provided via `statFun.custom`.}
#'   }
#'@param args A list of arguments required by the selected `statFun`. Specific arguments depend on the chosen method:
#'   \describe{
#'     \item{"lm"}{Requires `X` (design matrix), `y` (response variable). Optional: `Z` (covariates), `type`, `FL`, `getNull`, `n.perm`.}
#'     \item{"gam.mvwald"}{Requires `X`, `dat`, `gam.formula`, `lm.formula`, `y.in.gam`, `y.in.lm`.}
#'     \item{"gam.deltaRsq"}{Requires `X`, `dat`, `gam.full.formula`, `gam.null.formula`, `lm.formula`, `y.in.gam`, `y.in.lm`.}
#'     \item{"custom"}{Depends on the custom function implementation and should match the arguments specified in `statFun.custom`.}
#'   }
#'@param net.maps A list of vectors representing network maps. Each vector contains 0s (outside network/ROI) and 1s (inside network/ROI). The length of each vector must match the number of columns in `X`.
#'@param one.sided Logical. If `TRUE`, performs one-sided testing. Default is `TRUE`.
#'@param n.cores Integer. The number of cores to use for parallel processing. Default is `1`.
#'@param seed Integer. Random seed for reproducibility. Default is `NULL`.
#'@param what.to.return A character vector specifying the outputs to return. Options include:
#'   \describe{
#'     \item{"pval"}{P-values for each network (default).}
#'     \item{"ES"}{Enrichment scores (observed and null distributions).}
#'     \item{"ES.obs"}{Observed enrichment scores.}
#'     \item{"T.obs"}{Observed test statistics.}
#'     \item{"T.null"}{Null test statistics.}
#'     \item{"everything"}{All of the above outputs.}
#'   }
#'@param statFun.custom A user-defined function for custom statistical analysis. Required if `statFun` is set to `"custom"`. This function should accept arguments passed via `args`.
#'@return A list containing outputs specified in `what.to.return`. Default is p-values for each network.
#'@export
NEST = function(statFun, args, net.maps, one.sided = TRUE, n.cores = 1, seed = NULL, what.to.return = c("pval"), statFun.custom=NULL){

  # check input requirements:
  if (!is.list(net.maps)){
    message("net.maps argument should be a list, of vectors (one for each network)")
    return(NULL)
  }

  if (ncol(args$X) != length(net.maps[[1]])){
    message("each network should be specified as a vector with length equal to number of columns of X.")
    return(NULL)
  }

  if (!identical(as.numeric(sort(unique(unlist(net.maps)))), as.numeric(c(0,1)))){ # check that network maps are all 0s and 1s
    message("net.maps should include only 0's and 1's (1 = in network/ROI; 0 = outside network/ROI)")
    return(NULL)
  }

  # map brain-phenotype associations
  if (statFun == "lm"){
    required.args = c("X", "y")
    optional.args = c("Z", "type", "FL", "getNull","n.perm")

    args = checkArgs(args = args, required.args = required.args, optional.args = optional.args)
    if (!isFALSE(args)){
      statFun.out = statFun.lm(X = args$X,
                               y = args$y,
                               Z = args$Z,
                               type = args$type,
                               n.cores = n.cores,
                               seed = seed,
                               FL = args$FL,
                               n.perm = args$n.perm,
                               getNull = args$getNull)
    } else{
      message("fix args!")
      return(NULL)
    }

  }

  else if (statFun == "gam.mvwald"){

    # check args:
    required.args = c("X","dat","gam.formula","lm.formula","y.in.gam","y.in.lm")
    optional.args = c("n.perm")

    args = checkArgs(args = args, required.args = required.args, optional.args = optional.args)

    if (!isFALSE(args)){
      statFun.out = statFun.gam.mvwald(X = args$X,
                                      dat = args$dat,
                                      gam.formula = args$gam.formula,
                                      lm.formula = args$lm.formula,
                                      y.in.gam = args$y.in.gam,
                                      y.in.lm = args$y.in.lm,
                                      y.permute = args$y.permute,
                                      n.cores = n.cores, seed = seed,
                                      n.perm = args$n.perm,
                                      getNull = TRUE # if doing this inside NEST function, assume testing is being done (if just want to get map, could just use the statFun function directly)
                                      )
    }else{
      message("fix args!")
      return(NULL)
    }

  }

  else if (statFun == "gam.deltaRsq"){
    if (!isFALSE(args)){
      statFun.out = statFun.gam.deltaRsq(X = args$X,
                                      dat = args$dat,
                                      gam.full.formula = args$gam.full.formula,
                                      gam.null.formula = args$gam.null.formula,
                                      lm.formula = args$lm.formula,
                                      y.in.gam = args$y.in.gam,
                                      y.in.lm = args$y.in.lm,
                                      y.permute = args$y.permute,
                                      n.cores = n.cores, seed = seed,
                                      n.perm = args$n.perm,
                                      getNull = TRUE)
    }else{
      message("fix args!")
      return(NULL)
    }
  }

  # Custom function todo: test
  else if (statFun == "custom"){
    if(!is.null(statFun.custom)){
      statFun.out = do.call(statFun.custom, args)
    }else {
       message(("fix statFun.custom"))
       return(NULL)
    }
  }


  ES.list = mclapply(net.maps, FUN = function(net.map){
      ES.obs = enrichScore(stat.map = statFun.out$T.obs,
                  net.map = net.map,
                  one.sided = one.sided,
                  save.detail = F)

      ES.null = lapply(1:args$n.perm, FUN = function(k){
        enrichScore(stat.map = statFun.out$T.null[[k]],
                    net.map = net.map,
                    one.sided = one.sided,
                    save.detail = F)
      })

      return(list(ES.obs = ES.obs,
                  ES.null.dist = ES.null))
  }, mc.cores = n.cores)

  pval.list = mclapply(1:length(net.maps), FUN = function(net){
    pvalFun(obs = ES.list[[net]]$ES.obs, null.dist = ES.list[[net]]$ES.null.dist)
  })

  out = list()

  if ("everything" %in% what.to.return){
    out$pval = pval.list
    out$ES = ES.list
    out$T.obs = statFun.out$T.obs
    out$T.null = statFun.out$T.null
    return(out)
  }else{
    if ("pval" %in% what.to.return){ # default is to just return p-values
      out$pval = pval.list
    }

    if ("ES" %in% what.to.return){
      out$ES = ES.list
    }

    if ("ES.obs" %in% what.to.return){
      out$ES.obs = sapply(ES.list, FUN = function(x){x$ES.obs})
    }

    if ("T.obs" %in% what.to.return){
      out$T.obs = statFun.out$T.obs
    }

    if ("T.null" %in% what.to.return){
      out$T.null = statFun.out$T.null
    }
    return(out)
  }


}
